File: app/src/main/java/com/blog/demo/image/animation/LayoutAnimationActivity.java
package com.blog.demo.image.animation;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;

import com.blog.demo.R;

public class LayoutAnimationActivity extends Activity implements View.OnClickListener {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_image_animation_layout);
        findViewById(R.id.btn_start_animation).setOnClickListener(this);
        findViewById(R.id.btn_custom_layout_control).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btn_start_animation) {
            Animation animation = AnimationUtils.loadAnimation(this, R.anim.anim_in_from_right);
            LayoutAnimationController controller = new LayoutAnimationController(animation);
            controller.setDelay(0.1f);
            controller.setOrder(LayoutAnimationController.ORDER_NORMAL);

            ViewGroup container = findViewById(R.id.container);
            container.setLayoutAnimation(controller);
            container.startLayoutAnimation();
        } else if (v.getId() == R.id.btn_custom_layout_control) {
            Animation animation = AnimationUtils.loadAnimation(this, R.anim.anim_in_from_right);
            LayoutAnimationController controller = new CustomLayoutAnimationController(animation);
            controller.setDelay(0.1f);
            controller.setOrder(-1);

            ViewGroup container = findViewById(R.id.container);
            container.setLayoutAnimation(controller);
            container.startLayoutAnimation();
        }
    }

    private static class CustomLayoutAnimationController extends LayoutAnimationController {

        CustomLayoutAnimationController(Animation animation) {
            super(animation);
        }

        @Override
        protected int getTransformedIndex(AnimationParameters params) {
            if (getOrder() < 0) {
                int index = params.index;
                int count = (params.count + 1) / 2;

                if (index < count) {
                    return count - 1 - params.index;
                } else {
                    return params.index - count;
                }
            } else {
                return super.getTransformedIndex(params);
            }
        }
    }

}

File: app/src/main/java/com/blog/demo/image/animation/LayoutAnimationActivity.kt
package com.blog.demo.image.animation

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.view.animation.LayoutAnimationController
import android.widget.Button
import com.blog.demo.R

class LayoutAnimationActivity : Activity(), View.OnClickListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_image_animation_layout)

        findViewById<Button>(R.id.btn_start_animation).setOnClickListener(this)
        findViewById<Button>(R.id.btn_custom_layout_control).setOnClickListener(this)
    }

    override fun onClick(v: View) {
        if (v.id == R.id.btn_start_animation) {
            val animation = AnimationUtils.loadAnimation(this, R.anim.anim_in_from_right)
            val controller = LayoutAnimationController(animation)
            controller.delay = 0.1f
            controller.order = LayoutAnimationController.ORDER_NORMAL

            val container = findViewById<ViewGroup>(R.id.container)
            container.layoutAnimation = controller
            container.startLayoutAnimation()
        } else if (v.id == R.id.btn_custom_layout_control) {
            val animation = AnimationUtils.loadAnimation(this, R.anim.anim_in_from_right)
            val controller: LayoutAnimationController = CustomLayoutAnimationController(animation)
            controller.delay = 0.1f
            controller.order = -1

            val container = findViewById<ViewGroup>(R.id.container)
            container.layoutAnimation = controller
            container.startLayoutAnimation()
        }
    }

    private class CustomLayoutAnimationController constructor(animation: Animation) : LayoutAnimationController(animation) {

        override fun getTransformedIndex(params: AnimationParameters): Int {
            return if (order < 0) {
                val index = params.index
                val count = (params.count + 1) / 2
                if (index < count) {
                    count - 1 - params.index
                } else {
                    params.index - count
                }
            } else {
                super.getTransformedIndex(params)
            }
        }

    }

}